Introduce Dropzone Component

diogoredin opened the pull request:
# Description
This PR introduces a stateless Dropzone component using the [HTML5 Drag and Drop specification](https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API).

## Usage
```jsx
<Dropzone
  title="Upload PDF or Apple Wallet tickets"
  subtitle="If your original file contains multiple tickets make sure to upload all of them."
  action="Drop files here or click here to select"
  dropTitle="Release"
  accept={['.pdf', '.txt']}
  onFileChange={file => console.log(file)} // can set state here
  onUnacceptedFileChange={() => {}} // can do something in case its wrong
/>
```

## Changes

- [x] This has been done
- [ ] I still need to to this

## How to test

- Checkout this branch
- `$ yarn storybook`
- Look for `Dropzone`
- Test in different browsers / devices

## Screenshots
**Desktop**
<img width="683" alt="Screenshot 2022-08-23 at 17 17 01" src="https://user-images.githubusercontent.com/1096800/186196620-18fc0747-399b-4e94-a15a-3d3f1fd70136.png">

**Mobile**
<img width="445" alt="Screenshot 2022-09-05 at 14 24 13" src="https://user-images.githubusercontent.com/1096800/188468625-79d8a6d4-3cb3-4126-9051-a74e06d82d94.png">

## To be notified
@Julimpaiva @mon-d @TicketSwap/frontend 

## Links
[HG-2167](https://ticketswap.atlassian.net/browse/HG-2167)

diogoredin committed:
chore: Introduce test utils to test file uploads

This commit introduces two utils that allow us to test file uploads.
See the dataTransfer section here for more details:
https://testing-library.com/docs/dom-testing-library/api-events/

JIRA HG-2167

viktorrenkema commented:
I noticed some weird behaviour for the **restricted** variant:
- it doesn't seem to accept any pdf i **drop**, yet clicking the button and then uploading does work as it doesn't run onUnacceptedFileChange (as expected)
- however, I would expect a console log when I click the button in above variant, yet as visible in recording below, nothing shows.


https://user-images.githubusercontent.com/49148610/186352920-0bcb68b1-d76d-43bc-8f62-30d1a9802f2b.mov

Marruk commented:
Nice! I guess we should add this to Figma as well right @Julimpaiva?

Also, the screenshots don't look like the font is Proxima Nova, but maybe that's because it's in the Storybook environment.

diogoredin commented:
@viktorrenkema There was an issue in the condition not considering the extension for certain file names, it's fixed now 🙏 

<img width="1022" alt="Screenshot 2022-08-26 at 07 55 34" src="https://user-images.githubusercontent.com/1096800/186832209-35d8aeb2-83b4-4229-8dc5-30055d37ac7f.png">

@Marruk Yes, indeed, should be good once in use there I think.

RobVermeer commented on src/components/Dropzone/index.tsx:
```diff
@@ -0,0 +1,94 @@
+import React, { ChangeEvent, useRef } from 'react'
```
Let's create a `Dropzone.tsx` with all this and then inside `index.ts` export this component and all types that need to be exported (see `InputV2` as example)

diogoredin replied:
Applied 🙏

RobVermeer commented on src/components/Dropzone/index.tsx:
```diff
@@ -0,0 +1,94 @@
+  subtitle: string
+  action: string
+  ariaLabel?: string
+  accept?: Array<string>
```
`string[]` is what we use on most places

diogoredin replied:
oops, fixed!

RobVermeer commented on src/components/Dropzone/index.tsx:
```diff
@@ -0,0 +1,94 @@
+  action: string
+  ariaLabel?: string
+  accept?: Array<string>
+  onFileChange: Function
```
I think it is nice to type arguments and return value as well (if any). So more like this: `onFileChange: (event: SyntaticEvent) => void` (just an example, no idea wht this needs) (in anyway the arrow function definition is nicer that way imo, more explicit)

diogoredin replied:
Indeed, I added it as a placeholder and forgot to improve it 🙈 fixed it!

RobVermeer commented on src/components/Dropzone/index.tsx:
```diff
@@ -0,0 +1,94 @@
+  return (
+    <DropArea data-testid="droparea">
+      <>
+        <Illustration src={ILLUSTRATION} />
```
Images need width and height property on it, to calculate size and aspect ration before css is loaded (always the actual size of the image). Also an alt is needed, `alt=""` for images without meaning.

diogoredin replied:
Interesting! What happens if we don't do the calculation with CSS? Will it display the image in fullsize until the Javascript is loaded in the client side? 👀

diogoredin replied:
Fixed btw 🙏

diogoredin committed:
feat: Introduce Dropzone component with select file dialog

This commit introduce a basic Dropzone component with a button that allows
users to select a file from their operating system.

JIRA HG-2167

diogoredin committed:
feat: Improve responsive version of Dropzone

This commit modifies the Dropzone component so it displays a button
instead when viewed on mobile. This difference is introduced since the
most mobile devices don't enable drag and drop behavior.

JIRA HG-2167

diogoredin committed:
feat: Add capability of adding multiple files in Dropzone

This commit extends the Dropzone component so we are able to select more
than one file from the files dialog.

JIRA HG-2167

viktorrenkema commented on src/components/Dropzone/Dropzone.tsx:
```diff
@@ -0,0 +1,214 @@
+        onDrop={onDrop}
+      >
+        <img src={ILLUSTRATION} alt={dropTitle} />
+        <h2>{dropTitle}</h2>
```
Same as the h4 comment

diogoredin replied:
Looking at other components that aren't specifically made for holding text (e.g. `Card`) I think it's okay to make assumptions about the hierarchy on the page. But let me know if you think otherwise @TicketSwap/frontend I'm also okay with adding the prop.

viktorrenkema commented on src/components/Dropzone/Dropzone.tsx:
```diff
@@ -0,0 +1,214 @@
+  cursor: pointer;
+`
+
+const Title = styled(H4)`
```
This is now always an h4. Would it be an idea to accept a prop for the `as` attribute so the user of the Dropzone component can configure how this heading should be interpreted? Maybe an h4 doesn't make sense in all situations.

diogoredin replied:
Looking at other components that aren't specifically made for holding text (e.g. `Card`) I think it's okay to make assumptions about the hierarchy on the page. But let me know if you think otherwise @TicketSwap/frontend I'm also okay with adding the prop.

viktorrenkema commented on src/components/Dropzone/Dropzone.tsx:
```diff
@@ -0,0 +1,214 @@
+import { H4 } from '../Heading'
+import { SmallText } from '../Text'
+import { BaseButton } from '../BaseButton'
+import ILLUSTRATION from './files.svg'
```
I dont think the svg is centered 🤔 
![image](https://user-images.githubusercontent.com/49148610/188564281-05d389b0-f331-41fe-b12d-553616427bb4.png)

diogoredin replied:
Nice catch! Fixed 🙏

viktorrenkema commented:
I don't fully get the drop behaviour, sometimes i am able to drag the file anywhere in the overlay area (especially when I go 'around the sides' like in the video), but often it will also cancel the drop when I drag it to the same area from a different starting point.

I'd expect the entire overlay area to not cause a dismissal of the drop action.

https://user-images.githubusercontent.com/49148610/188565014-801e2076-3386-497e-a4d3-318680ebee9e.mov

Actually when typing this out, I figured it's the illustration and text that are not part of the drop area and therefore cancel the event. Working as expected when I remove those elements:

https://user-images.githubusercontent.com/49148610/188566632-a49149a6-4966-4934-96c7-20a6c122e7e2.mov

Maybe we can do something with z-indexes or other workarounds to have the drop-area be still 'on top' and not canceling the event?

diogoredin committed:
fix: Ensure whole dropzone overlay is draggable

This commit improves the dropzone overlay so the dragging behaviour does
not stop when the file is dragged over the files plus icon.

JIRA HG-2167

diogoredin commented:
@viktorrenkema The issue with the drop behaviour was solved, can you take another look? 🙏

viktorrenkema reviewed (approved):
🫡

RobVermeer commented:
:tada: This PR is included in version 27.2.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@ticketswap/solar/v/27.2.0)
- [GitHub release](https://github.com/TicketSwap/solar/releases/tag/v27.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
